### PR TITLE
185680669 client search improvement

### PR DESCRIPTION
**Pull request by ttoomey**
https://www.pivotaltracker.com/story/show/185680669

Note: I'm using raw sql in many places. If the logic here passes review I can convert it into arel if needed

* search against custom client names 
* match against middle initial or name, names that contain spaces in first/last name, and names with diacritics
* require at least one sequence of two or more chars to in the search term, max of 100 chars
* return search results ranked by distance from search term, weighting matches against the last name and the primary name more heavily
* refactored the block-passing interface to text search; this was necessary to be able to return results in ranked order
* add logging to RangeErrors rather than just ignoring them. What causes there errors, do they still occur?

performance impact:
* benchmarked against 150k client records locally on pg 12
* full text search from a cold DB is somewhat worse (from ~200ms to ~300ms). Without the GIN indexes this ends up being around 800ms. After the initial query, response time was more like 10ms 
* building indexes and generated columns did not take a noticeable amount over a db with ~150k client records (less than 10 seconds, I didn't benchmark it exactly)

misc:
* I did not change the hmis::client.client_search firstName/lastName methods. We could definitely do that
* we have a dedicated search_name_last field for that removes diacritics. I'm reusing search_name_full as a stand-in for first name when needed for prefix matching (since it begins with the first name and should work for prefix match). Perhaps we want a dedicated search_first_name to be more precise
* There's extra white space in the search_full_name field. This doesn't matter for the purposes of string similarity

### CODE REVIEW

File: drivers/client_access_control/app/controllers/client_access_control/clients_controller.rb [hunk 1]
```diff
@@ -55,7 +56,6 @@ def index
     end
 
     @clients = @clients.
-      distinct.
```
Comment (ttoomey) [inline drivers/client_access_control/app/controllers/client_access_control/clients_controller.rb:62]: This caused problems with sort rank ordering. Is it needed?

File: app/models/grda_warehouse/hud/client.rb [hunk 2]
```diff
@@ -1778,17 +1778,18 @@ def total_days_of_service
...
-          compact
-      rescue RangeError
-        return none
-      end
+    # @param client_scope [GrdaWarehouse::Hud::Client.source] find id matches in this scope
+    # @param sorted [Boolean] order results by closest match to text
+    def self.text_search(text, client_scope: nil, sorted: true)
+      result = (client_scope || self).text_searcher(text, sorted: sorted, resolve_for_join_query: true)
+      mapped = joins(%(JOIN "warehouse_clients" "warehouse_clients_x" ON "warehouse_clients_x"."destination_id" = "Client"."id" ))
+        .joins(%{JOIN (#{result.to_sql}) AS search_results ON search_results.client_id = "warehouse_clients_x".source_id})
```
Comment (ttoomey) [inline app/models/grda_warehouse/hud/client.rb]: if this construction seems okay, I can invest some time into turning this into arel. Also would be content to leave it as sql

File: app/models/grda_warehouse/hud/client.rb [hunk 3]
```diff
@@ -1778,17 +1778,23 @@ def total_days_of_service
...
-          where(where).
-          preload(:destination_client).
-          map { |m| m.destination_client&.id }.
-          compact
-      rescue RangeError
-        return none
-      end
+    # @param client_scope [GrdaWarehouse::Hud::Client.source] source clients to search in
+    # @param sorted [Boolean] order results by closest match to text
+    def self.text_search(text, client_scope: nil, sorted: true)
```
Comment (gigxz) [inline app/models/grda_warehouse/hud/client.rb]: Why is `client_scope` allowed to be `nil` now? I'd worry that this could lead to someone doing a search that is not scoped down by access, which it always should be.

Comment (gigxz) [reply app/models/grda_warehouse/hud/client.rb]: Nevermind.. I think thats ok, thats how lots of scopes work

File: drivers/hmis/app/models/hmis/hud/client.rb [hunk 4]
```diff
@@ -128,11 +119,11 @@ class << self
...
 
   scope :matching_search_term, ->(text_search) do
-    text_searcher(text_search, custom_name_options: CUSTOM_NAME_OPTIONS) do |where|
-      left_outer_joins(:names).where(where).pluck(:id)
-    rescue RangeError
-      return none
-    end
+    text_searcher(text_search, sorted: true)
+  rescue RangeError => e
+    # FIXME: what is this range error?
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/hud/client.rb]: I don't know tbh I think it was just copied from the warehouse code. @eanders do you know?

Comment (gigxz) [reply drivers/hmis/app/models/hmis/hud/client.rb]: Maybe if user searches for a really big number

File: app/models/concerns/client_search.rb [hunk 5]
```diff
@@ -89,19 +86,5 @@ def self.metaphone_search(where, field, text)
 
       where
     end
```
Comment (gigxz) [inline app/models/concerns/client_search.rb:94]: Was metaphone search completely removed? I think thats probably fine just checking.

**Comment by eanders** [comment]
Adding @awisema to this, I think you're more familiar with the previous version.

File: app/models/concerns/client_search.rb (hunk 5 above)
Comment (ttoomey) [reply app/models/concerns/client_search.rb:94]: Metaphone search should still work although maybe we should just replace it completely here. The current metaphone search isn't great for realtime data since the names are only built once a day.

Word similarity may be better in some cases (typos or non-western names) but it might be worse when the phonetic is different form the spelling (phu vs foo). With this construction we could support both metaphone and word similarity in the name search. We probably would add generated metaphone columns on the db rather than building uniqueNames in the app

File: app/models/grda_warehouse/hud/client.rb [hunk 6]
```diff
@@ -1778,17 +1778,23 @@ def total_days_of_service
...
+      # Get search results from client scope. Then return the unique destination client records that map to those matching source records
+      results = (client_scope || self).searchable.text_searcher(text, sorted: sorted, resolve_for_join_query: true)
+      grouped = GrdaWarehouse::WarehouseClient
+        .joins(%(JOIN (#{results.to_sql}) src_search_results ON "warehouse_clients"."source_id" = "src_search_results"."client_id"))
+        .select(Arel.sql(%("warehouse_clients"."destination_id" AS client_id, MAX(src_search_results.score) AS score)))
+        .group(Arel.sql('1'))
+
+      mapped = joins(%(JOIN (#{grouped.to_sql}) AS dst_search_results ON dst_search_results.client_id = "Client".id))
+      mapped = mapped.order(Arel.sql('dst_search_results.score DESC'), :id) if sorted
+      mapped
```
Comment (gigxz) [inline app/models/grda_warehouse/hud/client.rb]: This looks correct to me. I tested out searching by warehouse destination id and source client ID in the Warehouse and HMIS, both work as expected.

File: app/models/grda_warehouse/hud/client.rb [hunk 7]
```diff
@@ -1852,6 +1858,8 @@ def self.strict_search(criteria, client_scope: none)
           map { |m| m.destination_client.id }
         where(id: ids)
       rescue RangeError
+        # FIXME: what is this range error?
+        Sentry.capture_exception(e)
```
Comment (gigxz) [inline app/models/grda_warehouse/hud/client.rb]: I was able to hit this by searching for `99999999999999999999999999999999999999999999999`. I think we can remove this sentry log, and just add a code comment explaining

File: drivers/client_access_control/app/controllers/client_access_control/clients_controller.rb (hunk 1 above)
Comment (gigxz) [reply drivers/client_access_control/app/controllers/client_access_control/clients_controller.rb:62]: From what I can tell, no. Neither `strict_search` nor `text_search` should return duplicates, afaict.. I also tested searching for destination client that has multiple source clients, and it seems to be behaving correctly.

File: drivers/hmis/app/models/hmis/hud/client.rb (hunk 4 above)
Comment (gigxz) [reply drivers/hmis/app/models/hmis/hud/client.rb]: Confirmed that's what it is

File: app/models/concerns/client_search.rb [hunk 8]
```diff
@@ -44,36 +30,47 @@ def self.text_searcher(text, **kwargs, &block)
...
-        where = name_search(sa, :FirstName, query, **kwargs).
-          or(name_search(sa, :LastName, query, **kwargs))
+        matches_external_ids = false
+        if alpha_numeric && respond_to?(:search_by_external_id) && RailsDrivers.loaded.include?(:hmis_external_apis)
+          eid_t = HmisExternalApis::ExternalId.arel_table
+          external_id_scope = HmisExternalApis::ExternalId.for_clients
+            .where(eid_t[:value].eq(text))
+            .select(:source_id)
+          matches_external_ids = self.where(id: external_id_scope.select(:source_id)).any?
+        end
```
Comment (gigxz) [inline app/models/concerns/client_search.rb]: Why is this needed in addition to the `search_by_external_id` that is already called on line 51?

It doesnt seem like the right place to call it. If the external ID is numeric, or alphanumeric with length 32 or 36, this code wont get hit.

**Review by gigxz — COMMENTED** [review]
LGTM. One question around searching by ExternalID twice, seemingly.

File: app/models/grda_warehouse/hud/client.rb (hunk 7 above)
Comment (ttoomey) [reply app/models/grda_warehouse/hud/client.rb]: Thanks @gigxz! Looks like it's due to text_searcher trying to match by db pk, which is a 4 byte signed int, so any value greater than  2147483647 would trigger this exception. This seems a little funky. If an external ID exceeded this number it would be unsearchable.

I wonder if we need to search by Client.id at all? Internal db PKs aren't exposed to users afaik. @eanders  is that true?

Seems like could just make the following change to text_searcher and then remove all the range error rescues. 
```diff
-  where = sa[:PersonalID].eq(text).or(sa[:id].eq(text))
+ where = sa[:PersonalID].eq(text)
```

File: app/models/concerns/client_search.rb (hunk 8 above)
Comment (ttoomey) [reply app/models/concerns/client_search.rb]: So my thought was that it could be hard to combine ranked fuzzy name search with these other search conditions inside a single query, which is how the other conditions are handled. Perhaps there's an easy way to do that but it also could be a time sink.

To avoid dealing with it, we do an early return if there are no external id matches (at the expense of the extra query). Perhaps there's a better way to differentiate ID vs name search. Maybe if the term contains no numeric chars, we treat it as a name, otherwise it's a ID? Could also do the name search and return early if there are any results (assuming an ID would never fuzzy-match to a name).

> It doesnt seem like the right place to call it. If the external ID is numeric, or alphanumeric with length 32 or 36, this code wont get hit.

We shouldn't need to differentiate id vs name search if the term is numeric. However the alphanumeric string with 'magic' 32/36 lengths seems like it would be incorrect for searches against very long human names. That's the current behavior though.

Happy to huddle about this if needed.

File: app/models/grda_warehouse/hud/client.rb (hunk 7 above)
Comment (eanders) [reply app/models/grda_warehouse/hud/client.rb]: So, we use primary Client.id keys all the time, so yeah, these need to be searchable.  It's too bad they are only Int4.  I thought Rails usually made primary keys Int8, but something/someone early in the warehouse development set them to be Int4s on most tables.  We've fixed a bunch of other keys, but this one is pretty pervasive.  

Can we just rescue the exception and search the other items if we hit it?

Comment (ttoomey) [reply app/models/grda_warehouse/hud/client.rb]: Oh okay, good to know, thanks! In that case I think we can omit the the PK condition if the term would exceed the max for the column. I'll make that change. Right now it's all-or nothing

File: app/models/concerns/client_search.rb (hunk 8 above)
Comment (gigxz) [reply app/models/concerns/client_search.rb]: Ok, I think I get it. This extra query is here because we want to _avoid_ doing the NameSearch if the search term matches an external ID.

If the search query is numeric, the External ID query still occurs, it just happens down on line 55. 

Did I get that right?  Can we add a comment about why we're doing the search twice?

All of the External Ids we have right now are numeric, so `matches_external_ids` will never be true. But I can see the utility for the future, if we have some _alphanumeric_ external IDs to support.


Finally - do we need to repeat the actual query logic for searching by external ID? Can this be replaced with the below?

```
empty_where_clause = sa[:id].eq(nil)
matches_external_ids  = search_by_external_id(empty_where_clause, text).exists?
```


And - You're right, the 32/36 magic length would mess with searching for a name of that length. But to be considered "alphanumeric" there wouldn't be any spaces.. I think thats uncommon enough we should leave that behavior as-is.

Comment (ttoomey) [reply app/models/concerns/client_search.rb]: @gigxz Yes, that's exactly right, sorry this is confusing, I'll add some clarifying comments.

It's good to know the ids are only numeric right now, I was assuming UUIDs might be used too. So perhaps we could comment out the check for now for an easy performance gain.

I'll dry up the check use search_by_external_id. Confusingly, search_by_external_id returns a condition, not a scope, so the above change doesn't work as-is. I'll adjust it and incorporate it

Comment (gigxz) [reply app/models/concerns/client_search.rb]: Sounds good, thanks @ttoomey. Commenting it out for now seems like a good call.

**Comment by ttoomey** [comment]
I think all feedback has been addressed. Just needs final approval.

File: app/models/client_search_util/name_search.rb [hunk 9]
```diff
@@ -0,0 +1,152 @@
+module ClientSearchUtil
+  class NameSearch
```
Comment (eanders) [inline app/models/client_search_util/name_search.rb:2]:
```suggestion
  class NameSearch
    include ArelHelper
```

Comment (eanders) [reply app/models/client_search_util/name_search.rb:2]: Actually, never mind.  I was going to suggest we cleanup the SQL that's easy to cleanup into Arel, but I think everything here would be somewhat tricky to write.

File: app/models/concerns/client_search.rb [hunk 10]
```diff
@@ -17,63 +19,64 @@ def self.text_searcher(text, **kwargs, &block)
...
+          results = where(where)
         end
+      else
+        results = where(where)
       end
-      where(id: client_ids)
+
+      # we aren't dealing with a fuzzy string search here so we can't really rank the results
+      # return NULL as score as match is binary (either it matches an ID or it doesn't)
+      results = results.select(Arel.sql('"Client"."id" AS client_id'), Arel.sql('NULL AS score')) if resolve_for_join_query
```
Comment (eanders) [inline app/models/concerns/client_search.rb]: I think this one could be arel'd as:
```suggestion
      results = results.select(c_t[:id].as('client_id'), Arel.sql('NULL AS score')) if resolve_for_join_query
```

The table name quoting when the table case isn't lower case is one of the gotchas with cross-db use.

File: app/models/grda_warehouse/hud/client.rb [hunk 11]
```diff
@@ -397,7 +397,7 @@ class Client < Base
     end
 
     scope :full_text_search, ->(text) do
-      text_search(text, client_scope: current_scope)
+      text_search(text, sorted: false)
```
Comment (eanders) [inline app/models/grda_warehouse/hud/client.rb]: Oh, I missed that this dropped the `client_scope: current_scope` but text_search still uses it.  Let's add that back in.  (Unless there was a reason to remove it.)
```suggestion
      text_search(text, client_scope: current_scope, sorted: false)
```

**Review by eanders — APPROVED** [review]
@ttoomey thanks for all this!  It looks good.  I left a couple small comments inline, let me know if you want to discuss.

**Review by gigxz — APPROVED** [review]

File: app/models/grda_warehouse/hud/client.rb (hunk 11 above)
Comment (ttoomey) [reply app/models/grda_warehouse/hud/client.rb]: I took it out since current_scope is now the default for client_scope but it doesn't hurt to pass it in explicitly

**Comment by eanders** [comment]
Merging despite one failing test, I believe that's unrelated to this branch.
